tests/backward-compat-recovery-no-password/src/test/groovy/org/apache/bookkeeper/tests/TestCompatRecoveryNoPassword.groovy

```diff
@@ -0,0 +1,248 @@
+            if (rc == BKException.Code.OK) {
+                numSuccess.incrementAndGet()
+            }
+            latch.countDown()
+        }
+
+        long await() throws InterruptedException {
+            if (!latch.await(60, TimeUnit.SECONDS)) {
+                LOG.warn("Didn't get all responses in verification");
+                return 0;
+            } else {
+                return numSuccess.get();
+            }
+        }
+    }
+
+    private boolean verifyFullyReplicated(BookKeeper bookkeeper,
+                                          LedgerHandle lh,
+                                          long untilEntry) throws Exception {
+        LedgerMetadata md = getLedgerMetadata(bookkeeper, lh.getId())
+
+        def ensembles = md.getEnsembles()
+
+        def ranges = new HashMap<Long, Long>()
+        def keyList = new ArrayList(ensembles.keySet())
+        Collections.sort(keyList)
+        for (int i = 0; i < keyList.size() - 1; i++) {
+            ranges.put(keyList.get(i), keyList.get(i + 1))
+        }
+        ranges.put(keyList.get(keyList.size() - 1), untilEntry)
+
+        for (def e : ensembles.entrySet()) {
+            int quorum = md.getAckQuorumSize()
+            long startEntryId = e.getKey()
+            long endEntryId = ranges.get(startEntryId)
+            long expectedSuccess = quorum * (endEntryId - startEntryId)
+            int numRequests = e.getValue().size() * ((int) (endEntryId - startEntryId))
+
+            def cb = new ReplicationVerificationCallback(numRequests)
+            for (long i = startEntryId; i < endEntryId; i++) {
+                for (BookieSocketAddress addr : e.getValue()) {
+                    bookkeeper.getBookieClient()
+                        .readEntry(addr, lh.getId(), i, cb, addr)
+                }
+            }
+
+            long numSuccess = cb.await();
+            if (numSuccess < expectedSuccess) {
+                LOG.warn("Fragment not fully replicated ledgerId = {} startEntryId = {}"
+                         + " endEntryId = {} expectedSuccess = {} gotSuccess = {}",
+                         lh.getId(), startEntryId, endEntryId, expectedSuccess, numSuccess);
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Test that when we try to recover a ledger which doesn't have
+     * the password stored in the configuration, we don't succeed.
+     */
+    @Test
+    public void testRecoveryWithoutPasswordInMetadata() throws Exception {
+        int numEntries = 10
+        byte[] passwdCorrect = "AAAAAA".getBytes()
+        byte[] passwdBad = "BBBBBB".getBytes()
+
+        String currentVersion = System.getProperty("currentVersion")
+        String zookeeper = BookKeeperClusterUtils.zookeeperConnectString(docker)
+
+        BookKeeperClusterUtils.legacyMetadataFormat(docker)
+
+        // Create a 4.1.0 client, will update /ledgers/LAYOUT
+        def v410CL = MavenClassLoader.forBookKeeperVersion("4.1.0")
```

Comment:
I think we should try to close all classloaders

If I undersand correctly we are using forkCount = 0, so no new JVM will be spawned for test.
We should take care of not leaving memory (classloader) leaks

Reply:
Added a note to #1030